File: 0001-0050/0007/solution.py
class Solution:
    def reverse(self, x: int) -> int:
        INT_MIN, INT_MAX = -2 ** 31, 2 ** 31 - 1
        ans = 0
        while x:
            digit = x % 10
            if x < 0 and digit > 0:
                digit -= 10
            if x > 0 and ans > (INT_MAX - digit) // 10:
                return 0
            if x < 0 and ans < -((INT_MIN - digit) // -10):
                return 0
            ans = ans * 10 + digit
            x = x // 10 if x > 0 else -(x // -10)
        return ans


class Example:
    def __init__(self, x, ans):
        self.x = x
        self.ans = ans


def main():
    examples = [
        Example(123, 321),
        Example(-123, -321),
        Example(120, 21),
    ]

    solve = Solution()
    for i, example in enumerate(examples):
        ans = solve.reverse(example.x)
        if ans == example.ans:
            print(f"PASS: CASE {i}")
        else:
            print(f"FAIL: CASE {i}")


if __name__ == "__main__":
    main()

File: 0001-0050/0007/solution.cpp
#include <iostream>
#include <vector>

using namespace std;

class Solution {
public:
    int reverse(int x) {
        int ans = 0;
        while (x) {
            if (x > 0 && ans > (INT_MAX - x % 10) / 10) return 0;
            if (x < 0 && ans < (INT_MIN - x % 10) / 10) return 0;
            ans = ans * 10 + x % 10;
            x /= 10;
        }
        return ans;
    }
};

struct Example {
    int x;
    int ans;
};

int main() {
    vector<Example> examples = {
        {123, 321},
        {-123, -321},
        {120, 21},
    };
    
    Solution *solve = new Solution();
    for (int i = 0; i < examples.size(); i++) {
        int ans = solve->reverse(examples[i].x);
        if (ans == examples[i].ans) {
            cout << "PASS: CASE " << i << endl;
        } else {
            cout << "FAIL: CASE " << i << endl;
        }
    }
    
    return 0;
}